pkg/types/ccipocr3/generic_types.go

```diff
@@ -83,18 +83,26 @@ func (c ChainSelector) String() string {
 }
 
 type CCIPMsg struct {
-	ChainFeeLimit   BigInt        `json:"chainFeeLimit"`
-	Nonce           uint64        `json:"nonce"`
-	Sender          types.Account `json:"sender"`
-	Receiver        types.Account `json:"receiver"`
-	Strict          bool          `json:"strict"`
-	FeeToken        types.Account `json:"feeToken"`
-	FeeTokenAmount  BigInt        `json:"feeTokenAmount"`
-	Data            []byte        `json:"data"`
-	TokenAmounts    []TokenAmount `json:"tokenAmounts"`
-	SourceTokenData [][]byte      `json:"sourceTokenData"`
-
 	CCIPMsgBaseDetails
+	ChainFeeLimit   BigInt          `json:"chainFeeLimit"`
+	Nonce           uint64          `json:"nonce"`
+	Sender          types.Account   `json:"sender"`
+	Receiver        types.Account   `json:"receiver"`
+	Strict          bool            `json:"strict"`
+	FeeToken        types.Account   `json:"feeToken"`
+	FeeTokenAmount  BigInt          `json:"feeTokenAmount"`
+	Data            []byte          `json:"data"`
+	TokenAmounts    []TokenAmount   `json:"tokenAmounts"`
+	SourceTokenData [][]byte        `json:"sourceTokenData"`
+	Metadata        CCIPMsgMetadata `json:"metadata"`
+}
+
+type CCIPMsgMetadata struct {
+	// Version of the message metadata. Required in order to be able to parse the metadata
+	// by the underlying implementation.
+	Version uint64 `json:"version"`
+	// Data is the metadata payload. The underlying implementation should know how to parse this data.
+	Data []byte `json:"data"`
```

Comment:
What do we mean by "metadata payload" here? Where is this fetched from? Is this part of the onchain emitted message?

Reply:
It's populated by the underlying chain specific implementation e.g.
```
solMsg = CCIPMsgToSolana{
   sender
   recipient
   someSolSpecificField1
   someSolSpecificField2
}

ccipMsg = CCIPMsg{
   sender
   recipient
   []byte{ "someSolSpecificField1", "someSolSpecificField2" }
}
}

```

Reply:
I'm wondering if we should make CCIPMsgBaseDetails the only thing that is required by all chains and push the rest down to this `Data []byte` field

Reply:
That way the plugin is still able to perform its business logic; MsgHasher implementations will correctly decode the `Data` into the required structure to end up hashing

Reply:
Just synced offline. Will keep this as-is for now; suspect that we'll add to the `Data` field if nothing fits in the CCIPMsg explicit fields anyway